compiler/src/iree/compiler/Codegen/Dialect/PCF/IR/PCFBase.td

```diff
@@ -0,0 +1,280 @@
+// Base pcf attributes
+//===----------------------------------------------------------------------===//
+
+def PCF_DummyScopeAttr :
+    AttrDef<PCF_Dialect, "DummyScope", [
+      DeclareAttrInterfaceMethods<PCF_ScopeAttr>
+    ]> {
+  let mnemonic = "dummy_scope";
+  let summary = [{Dummy scope attribute used for testing }];
+  let description = [{
+    Scope that fails on all interface uses to use in testing where the scope
+    is not relevant.
+  }];
+  let parameters = (ins);
+  let assemblyFormat = [{}];
+
+  let extraClassDeclaration = [{
+    SmallVector<Value> getWorkerCounts(OpBuilder&, Location, int64_t) {
+      assert(false && "why are you here?");
+      return {};
+    }
+    SmallVector<Value> getWorkerIDs(OpBuilder&, Location, int64_t) {
+      assert(false && "why are you here?");
+      return {};
+    }
+  }];
+}
+
+def PCF_SequentialAttr :
+    AttrDef<PCF_Dialect, "Sequential", [
+      DeclareAttrInterfaceMethods<PCF_ScopeAttr>
+    ]> {
+  let mnemonic = "sequential";
+  let summary = [{Attribute representing sequential execution}];
+  let description = [{
+    Scope that reuses the current process as the sole executor of a parallel
+    region.
+  }];
+  let parameters = (ins);
+  let assemblyFormat = [{}];
+
+  let extraClassDeclaration = [{
+    SmallVector<Value> getWorkerCounts(OpBuilder& builder, Location loc, int64_t numIds) {
+      Value one = arith::ConstantIndexOp::create(builder, loc, 1);
+      SmallVector<Value> counts(numIds, one);
+      return counts;
+    }
+    SmallVector<Value> getWorkerIDs(OpBuilder& builder, Location loc, int64_t numIds) {
+      Value zero = arith::ConstantIndexOp::create(builder, loc, 0);
+      SmallVector<Value> ids(numIds, zero);
+      return ids;
+    }
+  }];
+}
+
+def PCF_SyncOnParentAttr :
+    AttrDef<PCF_Dialect, "SyncOnParent", [
+      DeclareAttrInterfaceMethods<PCF_SyncScopeAttr>
+    ]> {
+  let mnemonic = "sync_on_parent";
+  let summary = [{
+    Attribute indicating syncronization for the tied shape ref is dictated
+    solely by the parent.
+  }];
+  let description = [{
+    Attribute indicating that the shaped ref this attribute is tied to is only
+    fenced when the parent of the same scope returns.
+  }];
+  let parameters = (ins);
+  let assemblyFormat = [{}];
+
+  let extraClassDeclaration = [{
+    TypeRange getConcreteTypes(MLIRContext *context) {
+      // Expansion of this sync scope involves simply dropping it.
+      return TypeRange();
+    }
+    void enqueueWrite(OpBuilder& builder, ValueRange syncValues, Operation *writingOperation) {
+      // Nothing to do.
+      return;
+    }
+  }];
+}
+
+//===----------------------------------------------------------------------===//
+// Misc aliases
+//===----------------------------------------------------------------------===//
+
+def PCF_AnyRef: AnyTypeOf<[PCF_ShapedRef, AnyMemRef]>;
+def PCF_AnyShapedRef: AnyTypeOf<[PCF_ShapedRef]>;
```

Comment:
I think we might want to use the same constructions that AnyTensor / TensorOf and such out in upstream use.

Reply:
I couldn't for the life of me figure out how to do it the upstream way without it removing the `#dialect.mnemonic` prefix from all the type printers, so I just left it as is.